File: Solutions/1448-GoodNodesInTree/GoodNodesInTree.kt
class Solution {
    fun goodNodes(root: TreeNode?): Int {
        if (root == null) return 0
        return dfs(root, root.`val`)
    }

    fun dfs(
        node: TreeNode?,
        maxValueInPath: Int,
    ): Int {
        if (node == null) return 0

        val isNodeGood = node.`val` >= maxValueInPath
        val newMaxValueInPath = Math.max(node.`val`, maxValueInPath)

        val goodNodesLeft = dfs(node.left, newMaxValueInPath)
        val goodNodesRight = dfs(node.right, newMaxValueInPath)

        var numOfGoodNodes = goodNodesLeft + goodNodesRight
        if (isNodeGood) numOfGoodNodes++

        return numOfGoodNodes
    }
}

File: Solutions/All/1448-GoodNodesInTree/GoodNodesInTree.ts
import { TreeNode } from "../../../Types";

export function goodNodes(root: TreeNode | null): number {
  function dfs(node: TreeNode | null, maxValue: number): number {
    if (!node) return 0;

    const isNodeGood = node.val >= maxValue;
    maxValue = Math.max(node.val, maxValue);

    const goodNodesLeft = dfs(node.left, maxValue);
    const goodNodesRight = dfs(node.right, maxValue);

    let numOfGoodNodes = goodNodesLeft + goodNodesRight;
    if (isNodeGood) numOfGoodNodes++;

    return numOfGoodNodes;
  }

  if (!root) return 0;
  return dfs(root, root.val);
}
